Add logging for when inbuilt_inline_nn_modules will help with ID_MATCH guard triggered recompiles

ppanchalia opened the pull request:
We add a logging around when an ID_MATCH guard is added at a place where inbuilt_inline_nn_modules would inline it. This is done with the aim of tagging recompiles that could be avoided by setting inbuilt_inline_nn_modules flag.
It will help us log and track the flag's adoption and potentially quantify saving in the the number of recompiles.

Differential Revision: D80075975




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

anijain2305 commented on torch/_dynamo/convert_frame.py:
```diff
@@ -1078,8 +1078,19 @@ def count_args(code: CodeType) -> int:
+        # Recheck for recompilation, for when inline_inbuilt_nn_modules is set to False
+        if not config.inline_inbuilt_nn_modules:
+            reasons = get_and_maybe_log_recompilation_reasons(cache_entry, frame)
+            recompile_reason = (
```
Use some other variable than `recompile_reason` because this can overwrite the existing recompile_reason and you dont want that. In this condition, the only purpose is to identify a `inline_inbuilt_nn_module_candidate`

anijain2305 replied:
Its a fair question, but I would suggest not to do that in this PR. In this PR, it would be good if we don't change anything and just add more info. 

The main hesitation is that we do not qualify recompilation on a nn module instance as a True recompile today (for some technical reasons).

ppanchalia replied:
Sounds good, I'll fix this bit in a follow up diff

anijain2305 commented on torch/_dynamo/guards.py:
```diff
@@ -767,11 +767,22 @@ def get_verbose_code_part(code_part: str, guard: Optional[Guard]) -> str:
-    code_parts: Union[str, list[str]], guard: Optional[Guard]
+    code_parts: Union[str, list[str]],
+    guard: Optional[Guard],
+    recompile_hint: Optional[str] = False,
```
s/False/None - Optional[str] means either None or a string.

ppanchalia replied:
sorry, residue changes from draft.

anijain2305 commented on torch/_dynamo/guards.py:
```diff
@@ -4072,6 +4086,8 @@ def get_and_maybe_log_recompilation_reasons(
 
     code = frame.f_code
 
+    if skip_structured_traces:
```
In function `get_guard_fail_reason`, at L5040, we are appending into a global object `guard_failures`. We need to propagate `skip_structured_traces` to `get_guard_fail_reason` as well, and early return there. Maybe replace `skip_structured_traces` to `skip_logging` so that it makes sense for both the functions.

anijain2305 reviewed (commented):
Very close to be done. Can you also add a test? For test, you can add a guard hook and check if the str [inline-inbuilt-nn-modules-candidate] is present in the str representation of guards.  A pointer to do that is here - https://github.com/pytorch/pytorch/blob/2898d3f965e5cd9d02fc2ecdab7c580fd457fea9/test/dynamo/test_guard_manager.py#L1055

anijain2305 commented on test/dynamo/test_recompiles.py:
```diff
@@ -4,9 +4,57 @@
 
 class RecompileTests(torch._dynamo.test_case.TestCase):
+    def test_inline_inbuilt_nn_modules_candidate(self):
+        self.maxDiff = None
```
Unused

anijain2305 commented on torch/_dynamo/convert_frame.py:
```diff
@@ -1080,6 +1080,28 @@ def count_args(code: CodeType) -> int:
             )
+        # Recheck for recompilation, for when inline_inbuilt_nn_modules is set to False
+        inline_inbuilt_nn_modules_candidate = False
+        if not config.inline_inbuilt_nn_modules and frame:
```
Why `and frame`?

anijain2305 reviewed (commented):
Test is failing. But the code looks good.

ppanchalia committed:
Add logging for when inbuilt_inline_nn_modules will help with ID_MATCH guard triggered recompiles (#160592)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/160592

We add a logging around when an ID_MATCH guard is added at a place where inbuilt_inline_nn_modules would inline it. This is done with the aim of tagging recompiles that could be avoided by setting inbuilt_inline_nn_modules flag.

It will help us log and track the flag's adoption and potentially quantify saving in the the number of recompiles.

Test Plan:
rm -rf /tmp/r && TORCH_TRACE=/tmp/r buck2 run //scripts/ppanchalia/pt2_workspace:inline_inbuilt_nn_modules_test and CI Tests

And UT added

Rollback Plan:

Differential Revision: D80075975

ppanchalia commented:
@pytorchbot merge
